doc/server_reflection_tutorial.md

```diff
@@ -0,0 +1,190 @@
+# gRPC Server Reflection Tutorial
+
+gRPC Server Reflection provides information about publicly-accessible gRPC
+services on a server, and assists clients in runtime to construct RPC
+requests/responses without precompiled service information. It has been
+supported by gRPC CLI, which could be used to introspect server protos and
+send/receive test RPCs.
+
+## Eanble Server Reflection
+
+### Enable server reflection in C++ servers
+
+C++ Server Reflection is offered as an add-on library, `libgrpc++_reflction`.
+To enable C++ server reflection, you can simply link this library to your server
+binary.
+
+Some platforms (e.g. Ubuntu 11.10 onwards) only link in libraries that directly
+contain symbols used by the application. On these platforms, LD flag
+`--no-as-needed` is needed for for dynamic linking and `--whole-archive` is
+needed for for static linking.
+
+Here is an [example](examples/cpp/helloworld/Makefile#L37#L45) for enabling c++
+server reflection on Linux and MacOS.
+
+## Test services using Server Reflection
+
+gRPC Server Reflection has been supported by gRPC CLI. After enabling Server
+Reflection in a server application, you can use gRPC CLI to test its services.
+
+Instructions on how to use gRPC CLI can be found at
+[command_line_tool.md](doc/command_line_tool.md), or using `grpc_cli help`
+command.
+
+Here we use `examples/cpp/helloworld` as an example to show the use of gRPC
+Server Reflection and gRPC CLI. First, we need to build gRPC CLI and setup an
+example server with Server Reflection enabled.
+
+- Setup an example server
+
+  Server Reflection has already been enabled in the
+  [Makefile](examples/cpp/helloworld/Makefile) of the helloworld example. We can
+  simply make it and run the greeter_server.
+
+  ```sh
+  $ make -C examples/cpp/helloworld
+  $ examples/cpp/helloworld/greeter_server &
+  ```
+
+- Build gRPC CLI
+
+  ```sh
+  make grpc_cli
+  ```
+
+### List services
+
+`grpc_cli ls` command can list services and methods exposed at a given port
+
+- List all the services exposed at a given port
+
+  ```sh
+  $ bins/opt/grpc_cli ls localhost:50051
```

Comment:
It might not be good to add `bins/opt/` to $PATH, as this folder contains all the test binaries and will be deleted by `make clean`. Once the command line tool is mature enough, we can consider to move its binary to grpc's root folder or install it with `make install`. Before this, we may recommend users to `cd` into `bins/opt/`, and the prefix will no longer be needed in the rest of the tutorial.